x/mongo/driver/auth/creds/awscreds.go

```diff
@@ -0,0 +1,187 @@
+		return &awsv4.StaticProvider{Value: awsv4.Value{
+			AccessKeyID:     ac.Username,
+			SecretAccessKey: ac.Password,
+			SessionToken:    ac.Token,
+		}}, nil
+	}
+	return nil, nil
+}
+
+// AwsCredentialProvider provides AWS credentials.
+type AwsCredentialProvider struct {
+	HTTPClient *http.Client
+}
+
+func executeAWSHTTPRequest(ctx context.Context, httpClient *http.Client, req *http.Request) ([]byte, error) {
+	ctx, cancel := context.WithTimeout(ctx, defaultHTTPTimeout)
+	defer cancel()
+	resp, err := httpClient.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	return ioutil.ReadAll(resp.Body)
+}
+
+// getEC2Credentials generates EC2 credentials.
+func (p *AwsCredentialProvider) getEC2Credentials(ctx context.Context) (*awsv4.StaticProvider, error) {
+	// get token
+	req, err := http.NewRequest("PUT", awsEC2URI+awsEC2TokenPath, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("X-aws-ec2-metadata-token-ttl-seconds", "30")
+
+	token, err := executeAWSHTTPRequest(ctx, p.HTTPClient, req)
+	if err != nil {
+		return nil, err
+	}
+	if len(token) == 0 {
+		return nil, errors.New("unable to retrieve token from EC2 metadata")
+	}
+	tokenStr := string(token)
+
+	// get role name
+	req, err = http.NewRequest("GET", awsEC2URI+awsEC2RolePath, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("X-aws-ec2-metadata-token", tokenStr)
+
+	role, err := executeAWSHTTPRequest(ctx, p.HTTPClient, req)
+	if err != nil {
+		return nil, err
+	}
+	if len(role) == 0 {
+		return nil, errors.New("unable to retrieve role_name from EC2 metadata")
+	}
+
+	// get credentials
+	pathWithRole := awsEC2URI + awsEC2RolePath + string(role)
+	req, err = http.NewRequest("GET", pathWithRole, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("X-aws-ec2-metadata-token", tokenStr)
+	creds, err := executeAWSHTTPRequest(ctx, p.HTTPClient, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var es2Resp ecsResponse
+	err = json.Unmarshal(creds, &es2Resp)
+	if err != nil {
+		return nil, err
+	}
+	return AwsCredentials{
+		Username: es2Resp.AccessKeyID,
+		Password: es2Resp.SecretAccessKey,
+		Token:    es2Resp.Token,
+	}.ValidateAndMakeCredentials()
+}
+
+// GetCredentials generates AWS credentials.
+func (p *AwsCredentialProvider) GetCredentials(ctx context.Context) (*awsv4.StaticProvider, error) {
+	// Credentials from environment variables
+	ac := AwsCredentials{
+		Username: os.Getenv("AWS_ACCESS_KEY_ID"),
+		Password: os.Getenv("AWS_SECRET_ACCESS_KEY"),
+		Token:    os.Getenv("AWS_SESSION_TOKEN"),
+	}
+
+	creds, err := ac.ValidateAndMakeCredentials()
+	if creds != nil || err != nil {
+		return creds, err
+	}
+
+	// Credentials from ECS metadata
+	relativeEcsURI := os.Getenv("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI")
+	if len(relativeEcsURI) > 0 {
+		fullURI := awsRelativeURI + relativeEcsURI
+
+		req, err := http.NewRequest("GET", fullURI, nil)
```

Comment:
```suggestion
		req, err := http.NewRequest(http.MethodGet, fullURI, nil)
```

Suggest using the [HTTP constants](https://pkg.go.dev/net/http#MethodPut).